### PR TITLE
fix: make provider batch apply and undo crash-safe

### PR DESCRIPTION
**Pull request by Brisbanehuang**
## What changed

- persist provider batch apply claims and results in PostgreSQL with a 24-hour retention window
- enforce the complete preview preimage with row lock before applying any provider update
- make preview tokens single-consumption and idempotency keys durable across processes and restarts
- keep sensitive proxy URLs out of durable undo payloads
- make provider undo atomic: provider restoration and undo-token consumption commit in the same transaction
- add unit coverage and a PostgreSQL integration suite for concurrency, rollback, replay, and undo behavior

## Related

- Follow-up to #806 — hardens the provider batch preview/apply/undo pipeline and in-memory undo snapshot store introduced there, replacing them with a durable PostgreSQL ledger.

## Why

The existing preview/apply flow could validate against fresh state outside the write transaction. Concurrent provider edits or repeated execution of the same preview could therefore overwrite newer values or perform the same approved operation more than once.

The previous undo flow also consumed its token before restoring provider rows, leaving a crash window where the token was permanently consumed but the provider changes were not reverted.

## Behavior

- any drift in provider type, enabled state, or a field touched by the approved patch aborts the complete batch with no writes
- exclusions reduce only the write set; they do not reduce the preimage validation set
- retries with the same key, preview, and payload replay the committed result
- reuse of a key with another payload conflicts, and a preview cannot be consumed under another key
- undo row restoration and durable token consumption are one PostgreSQL transaction
- cache and circuit-breaker invalidation remain idempotent post-commit effects

## Validation

- `bun run typecheck`
- 62 focused provider batch unit tests
- PostgreSQL 16 integration suite covering concurrent claims, preview single-consumption, preimage rollback, SQL rollback, durable replay, atomic undo, and failed-undo rollback
- changed-file Biome check and `git diff --check`

<details open><summary><h3>Greptile Summary</h3></summary>

This PR hardens the provider batch preview/apply/undo pipeline by replacing the previous in-memory undo snapshot store with a durable PostgreSQL ledger, making the flow resilient to process crashes and concurrent edits.

- **Atomic apply**: a single transaction claims the preview token via INSERT, acquires row-level locks on all provider rows in a deterministic order, validates the complete preimage (including excluded providers), writes the effective updates, and transitions the ledger row from `applying` to `applied` — all or nothing.
- **Atomic undo**: a single transaction locks the ledger row with `FOR UPDATE`, applies provider restoration, and marks the undo token consumed in the same commit, eliminating the prior crash window.
- **Idempotency and replay**: duplicate apply requests with the same `claimKey`/`previewToken`/payload replay the committed result from the ledger; sensitive proxy URLs are stripped from the durable undo preimage; expired-row cleanup is moved outside the critical transaction to avoid blocking provider locks on large backlogs.
</details>

<details open><summary><h3>Confidence Score: 5/5</h3></summary>

The core transactional correctness is solid — provider locks are acquired in a consistent order, the preimage CAS check is complete, and undo token consumption is committed atomically with provider restoration. The two findings are quality concerns rather than functional defects on any live path.

The apply and undo transactions are both correctly structured, the shared SENSITIVE_PROVIDER_BATCH_UNDO_KEYS constant resolves the prior duplicate-literal drift risk, and all new repository status values are properly handled in the action layer. The exported-but-unused consumeProviderBatchUndo is a cleanup concern, not a runtime defect; the missing DB-level CHECK on the status column is a defence-in-depth gap, not a broken invariant.

src/repository/provider.ts — the exported consumeProviderBatchUndo function should be reviewed for removal. src/drizzle/schema.ts — the status column would benefit from a CHECK constraint to match the project's use of pgEnum elsewhere.
</details>

<details open><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| src/repository/provider.ts | Core ledger logic: adds applyProviderBatchOperationIfUnchanged, undoProviderBatchOperation, findProviderBatchApplyOperation, and supporting helpers. The atomic apply and undo transactions are both correctly structured. consumeProviderBatchUndo is exported but never called — a footgun that should be removed. |
| src/actions/providers.ts | Action layer wiring: applyProviderBatchPatch now builds the durable undo preimage (stripping sensitive keys via the shared SENSITIVE_PROVIDER_BATCH_UNDO_KEYS constant) and delegates to applyProviderBatchOperationIfUnchanged; undoProviderPatch falls back to durable ledger when in-memory snapshot is absent and routes to the atomic undoProviderBatchOperation. All new status values are correctly handled. |
| src/drizzle/schema.ts | New provider_batch_apply_operations table: correct use of unique indexes for preview_token, operation_id, and undo_token; status column uses varchar(32) with TypeScript-level typing but no DB CHECK constraint, inconsistent with other enum columns that use pgEnum. |
| drizzle/0109_nice_shriek.sql | Migration creates provider_batch_apply_operations with four indexes (unique on preview_token, operation_id, undo_token; btree on expires_at). DDL is consistent with the Drizzle schema definition. |
| src/lib/provider-batch-patch-error-codes.ts | Adds UNDO_STALE error code and moves SENSITIVE_PROVIDER_BATCH_UNDO_KEYS here as the single source of truth, resolving the previous duplicate-literal concern. |
| src/app/api/v1/resources/providers/handlers.ts | Maps UNDO_STALE to HTTP 409 alongside the other conflict codes; no other handler changes. |

</details>

<details open><summary><h3>Sequence Diagram</h3></summary>

```mermaid
sequenceDiagram
    participant C as Client
    participant A as Action Layer
    participant L as Ledger (PG)
    participant P as Providers (PG)

    Note over C,P: Apply path
    C->>A: applyProviderBatchPatch(claimKey, previewToken, payload)
    A->>L: findProviderBatchApplyOperation
    alt replay or conflict
        L-->>A: "replay | idempotency_conflict | preview_consumed"
        A-->>C: cached result or 409
    else not_found
        A->>L: "BEGIN TX INSERT ledger status=applying"
        L-->>A: inserted
        A->>P: SELECT FOR UPDATE ORDER BY id
        A->>P: validate preimage CAS
        A->>P: UPDATE effective providers
        A->>L: "UPDATE ledger status=applied"
        A->>L: COMMIT
        A-->>C: 200 operationId undoToken
    end

    Note over C,P: Undo path (durable)
    C->>A: undoProviderPatch(undoToken, operationId)
    A->>L: findProviderBatchUndoOperation
    A->>L: BEGIN TX SELECT FOR UPDATE ledger row
    A->>P: UPDATE providers to preimage
    A->>L: "UPDATE ledger undo_consumed_at=now"
    A->>L: COMMIT
    A-->>C: 200 revertedCount
```
</details>

<sub>Reviews (5): Last reviewed commit: ["Merge dev into provider batch safety"](https://github.com/ding113/claude-code-hub/commit/71b7a8ebadfe182b90ad409321db3b283638848a) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=44315693)</sub>

### CODE REVIEW

File: src/actions/providers.ts [hunk 1]
```diff
@@ -1965,6 +2018,148 @@ function generatePreviewRows(
...
+        forceCloseCircuitState(providerId, { reason: "circuit_breaker_disabled" }).catch(
+          (error) => {
+            logger.warn("applyProviderBatchPatch:force_close_circuit_failed", {
+              providerId,
+              error: error instanceof Error ? error.message : String(error),
+            });
+          }
+        )
+      )
+    );
```
[the remark this answers is not in the document]
Comment (Brisbanehuang) [reply src/actions/providers.ts:2158]: Fixed in cf74ff85. Before replaying `forceCloseCircuitState`, the post-commit handler reloads each provider and only closes the circuit when the current failure threshold is still disabled.

File: src/repository/provider.ts [hunk 2]
```diff
@@ -1300,89 +1391,630 @@ export async function updateProvidersBatch(
...
+  if (
+    current !== null &&
+    expected !== null &&
+    typeof current === "object" &&
+    typeof expected === "object"
+  ) {
+    return JSON.stringify(current) === JSON.stringify(expected);
+  }
+  return false;
+}
```
[the remark this answers is not in the document]
Comment (Brisbanehuang) [reply src/repository/provider.ts:1469]: Fixed in cf74ff85. The preimage comparator now performs recursive structural equality for arrays and objects, so JSONB object key order no longer causes a false stale result. Typecheck and the focused provider tests pass.

File: tests/unit/actions/providers-patch-actions-contract.test.ts [hunk 3]
```diff
@@ -5,7 +5,19 @@ import { buildRedisMock, createRedisStore } from "./redis-mock-utils";
...
 const { store: redisStore, mocks: redisMocks } = createRedisStore();
+const applyLedger = new Map<
+  string,
+  {
+    previewToken: string;
+    payloadFingerprint: string;
+    result: Record<string, unknown>;
+    undoAvailable: boolean;
+  }
+>();
```
[the remark this answers is not in the document]
Comment (Brisbanehuang) [reply tests/unit/actions/providers-patch-actions-contract.test.ts:22]: Fixed in cf74ff85. Both action tests now use the concrete `ProviderBatchApplyLedgerResult` type from the schema instead of `Record<string, unknown>`, preserving typed access to `applyResult.undoToken` and `operationId`.

File: tests/unit/actions/providers-apply-engine.test.ts [hunk 4]
```diff
@@ -1,11 +1,23 @@
...
 import { PROVIDER_BATCH_PATCH_ERROR_CODES } from "@/lib/provider-batch-patch-error-codes";
 
 const getSessionMock = vi.fn();
 const findAllProvidersFreshMock = vi.fn();
 const updateProvidersBatchMock = vi.fn();
+const findProviderBatchApplyOperationMock = vi.fn();
+const applyProviderBatchOperationIfUnchangedMock = vi.fn();
+const consumeProviderBatchUndoMock = vi.fn();
 const publishCacheInvalidationMock = vi.fn();
 const redisStore = new Map<string, { value: string; expiresAt: number }>();
```
[the remark this answers is not in the document]
Comment (Brisbanehuang) [reply tests/unit/actions/providers-apply-engine.test.ts:11]: Verified against the current branch: there is only one module-level `redisStore` declaration in this test. `bun run typecheck` and the focused test suite pass, so no code change was needed for this finding.

File: src/repository/provider.ts [hunk 5]
```diff
@@ -1300,89 +1391,630 @@ export async function updateProvidersBatch(
...
+ * reconciliation so a prior post-commit failure can heal on retry.
+ */
+export async function findProviderBatchApplyOperation(
+  input: FindProviderBatchApplyOperationInput
+): Promise<ProviderBatchApplyOperationLookupResult> {
+  const now = new Date();
+  await db
+    .delete(providerBatchApplyOperations)
+    .where(lt(providerBatchApplyOperations.expiresAt, now));
+  const rows = await db
```
[the remark this answers is not in the document]
Comment (Brisbanehuang) [reply src/repository/provider.ts:1590]: Fixed in cf74ff85. The lookup remains read-only and now filters `expiresAt > now` in SQL. Expired rows are no longer replayed, while cleanup is handled outside the provider-locking transaction on the apply path.

File: src/actions/providers.ts [hunk 6]
```diff
@@ -1965,6 +2024,151 @@ function generatePreviewRows(
...
+  operation: ProviderBatchApplyLedgerResult,
+  undoAvailable: boolean,
+  initialSnapshot?: Pick<ProviderPatchUndoSnapshot, "providerIds" | "preimage">
+): Promise<void> {
+  if (!undoAvailable) {
+    return;
+  }
+  const remainingTtlSeconds = Math.floor(
+    (new Date(operation.applyResult.undoExpiresAt).getTime() - Date.now()) / 1000
+  );
```
[the remark this answers is not in the document]
Comment (Brisbanehuang) [reply src/actions/providers.ts]: Fixed in d81bb210. The remaining Redis TTL now uses Math.ceil, so the snapshot cannot expire before the advertised undoExpiresAt timestamp. I also moved the replay regression test to a 1 ms boundary and assert that the Redis expiry is at or after the advertised expiry (and less than one second later), which fails with the previous Math.floor behavior.

File: tests/integration/provider-batch-ledger-postgres.test.ts [hunk 7]
```diff
@@ -0,0 +1,374 @@
...
+      undoProviderBatchOperation({
+        undoToken: applied.result.applyResult.undoToken,
+        operationId: applied.result.applyResult.operationId,
+        groups: [
+          { ids: [providerId], updates: { priority: 1 } },
+          { ids: [providerId + 999_999], updates: { priority: 1 } },
+        ],
+        revertedAt: new Date(),
+      })
+    ).rejects.toThrow();
```
[the remark this answers is not in the document]
Comment (Brisbanehuang) [reply tests/integration/provider-batch-ledger-postgres.test.ts]: Fixed in d81bb210. The PostgreSQL integration test now expects the repository API result { status: "mismatch" } instead of a rejected promise. The subsequent assertions still verify transaction rollback and undoAvailable=true. I ran this suite against a temporary PostgreSQL 16 database: 7/7 tests passed.

File: src/actions/providers.ts [hunk 8]
```diff
@@ -2311,14 +2543,65 @@ export async function undoProviderPatch(
...
+          operationId: parsed.data.operationId,
+          groups: [...preimageGroups.values()],
+          revertedAt: new Date(nowMs),
+        })
+      : {
+          status: "reverted" as const,
+          revertedCount: await (async () => {
+            let count = 0;
+            for (const { ids, updates } of preimageGroups.values()) {
+              count += await updateProvidersBatch(ids, updates);
```
[the remark this answers is not in the document]
Comment (Brisbanehuang) [reply src/actions/providers.ts:2578]: Fixed in e0db39a9. Volatile undo tokens are now atomically consumed with Redis GET+DEL before any provider write. If the token was already consumed or Redis consumption fails, undo returns UNDO_EXPIRED and does not touch the database. The post-write Redis cleanup remains only for durable snapshots. I also added regressions for consume-before-write ordering, Redis failure, and two concurrent undo calls (exactly one succeeds and updateProvidersBatch runs once). Verified 53/53 focused undo/ledger tests, typecheck, and the Biome 2.5.4 full-repository format check.

File: tests/unit/actions/providers-apply-engine.test.ts [hunk 9]
```diff
@@ -52,6 +62,9 @@ vi.mock("@/lib/auth", () => ({
 vi.mock("@/repository/provider", () => ({
   findAllProvidersFresh: findAllProvidersFreshMock,
   updateProvidersBatch: updateProvidersBatchMock,
+  findProviderBatchApplyOperation: findProviderBatchApplyOperationMock,
+  applyProviderBatchOperationIfUnchanged: applyProviderBatchOperationIfUnchangedMock,
+  consumeProviderBatchUndo: consumeProviderBatchUndoMock,
   deleteProvidersBatch: vi.fn(),
```
[the remark this answers is not in the document]
Comment (Brisbanehuang) [reply tests/unit/actions/providers-apply-engine.test.ts:68]: I could not reproduce the reported import failure. On the reviewed e0db39a9 head, running `bunx vitest run tests/unit/actions/providers-apply-engine.test.ts` without those two mock exports passes 21/21, and the GitHub full Unit Tests check also passed. Vitest partial mocks only need exports exercised by this suite; this file tests apply paths and never calls the undo ledger functions. The dedicated undo suites already mock `findProviderBatchUndoOperation` and `undoProviderBatchOperation`. Adding unused mocks here would not add coverage or change behavior, so I am resolving this as a false positive with the direct test evidence above.

File: src/repository/provider.ts [hunk 10]
```diff
@@ -1300,89 +1392,698 @@ export async function updateProvidersBatch(
...
+      }
+
+      let revertedCount = 0;
+      for (const group of input.groups) {
+        const ids = [...new Set(group.ids)].sort((a, b) => a - b);
+        if (ids.length === 0) continue;
+        const updated = await tx
+          .update(providers)
+          .set(buildProviderBatchSetClauses(group.updates, input.revertedAt))
+          .where(and(inArray(providers.id, ids), isNull(providers.deletedAt)))
```
[the remark this answers is not in the document]
Comment (Brisbanehuang) [reply src/repository/provider.ts:1956]: Fixed in a8527172. After the durable undo transaction commits successfully, undoProviderBatchOperation now reuses the existing enabled-provider endpoint reconciliation for every group that restores isEnabled=true. Conflict, expired, and rolled-back mismatch paths do not run the post-commit reconciliation. I added a real PostgreSQL 16 integration regression that disables an enabled provider, confirms the endpoint is absent, performs durable undo, and verifies both the provider and an active non-deleted endpoint are restored. TDD result: the new test failed with 0 endpoints before the fix; afterward the PostgreSQL suite passes 8/8. Related unit tests pass 40/40, plus typecheck and Biome 2.5.4 full-repository format check.
